File: bindings/python/perf/multi/perf_multi_router_router_client.py
import sys
import time
from contextlib import ExitStack

import zlink

from perf_multi_common import (
    apply_multi_auto_hwm_msg_unit,
    apply_multi_socket_options,
    benchmark_run_id,
    configure_multi_tls_client,
    is_active_message,
    latency_ns_from_message,
    new_payload,
    parse_client_args,
    perf_client_context,
    print_result_lines,
    received_metric_payload,
    recv_nonblocking,
    resolve_multi_connect_ready_timeout_ms,
    result_metrics,
    safe_poll,
    send_nonblocking,
    stamp_payload,
    wait_monitor_event,
)


def main(argv=None):
    args = parse_client_args(argv or sys.argv[1:], pattern="router_router")
    run_id = benchmark_run_id()
    payloads = [new_payload(args.msg_size) for _ in range(args.clients)]
    # C run_echo_window_round_robin: purely deadline-driven, no stop token.
    waiting_reply = [False] * args.clients
    send_pending = [True] * args.clients
    received = 0
    latencies = []
    seq = 0

    with perf_client_context() as ctx:
        apply_multi_auto_hwm_msg_unit(ctx, args.msg_size)
        sockets = [zlink.create_router_socket(ctx) for _ in range(args.clients)]
        try:
            with ExitStack() as stack:
                monitors = []
                for index, sock in enumerate(sockets):
                    sock.set_routing_id(f"CLIENT-{index}".encode("ascii"))
                    sock.router_options.connect_routing_id = b"SERVER"
                    monitor = stack.enter_context(
                        sock.monitor_open(zlink.MonitorEventMask.CONNECTION_READY)
                    )
                    configure_multi_tls_client(sock, args.transport)
                    apply_multi_socket_options(sock)
                    sock.connect(args.endpoint)
                    monitors.append(monitor)
                for monitor in monitors:
                    wait_monitor_event(
                        monitor,
                        zlink.MonitorEventMask.CONNECTION_READY,
                        timeout_ms=resolve_multi_connect_ready_timeout_ms(),
                    )

                active_deadline = time.perf_counter() + args.duration
                recv_storage = [zlink.create_received() for _ in sockets]
                with zlink.create_poller() as poller:
                    poll_events = zlink.create_poll_events(max(1, len(sockets)))
                    for index, sock in enumerate(sockets):
                        poller.add_socket(
                            sock,
                            zlink.PollEventFlag.POLLIN | zlink.PollEventFlag.POLLOUT,
                            index,
                        )
                    # C run_echo_window_round_robin: deadline-driven (no stop
                    # token); -1 signal-driven poll dispatching only ready
                    # sources.
                    while time.perf_counter() < active_deadline:
                        for index, current_sock in enumerate(sockets):
                            if waiting_reply[index] or not send_pending[index]:
                                continue
                            seq += 1
                            payload = stamp_payload(
                                payloads[index],
                                phase=1,
                                run_id=run_id,
                                seq=seq,
                            )
                            if send_nonblocking(
                                current_sock,
                                payload,
                                routing_id=b"SERVER",
                            ):
                                waiting_reply[index] = True
                                send_pending[index] = False

                        interest = [
                            i
                            for i in range(len(sockets))
                            if waiting_reply[i] or send_pending[i]
                        ]
                        if not interest:
                            for i in range(len(sockets)):
                                if not waiting_reply[i]:
                                    send_pending[i] = True
                            continue

                        remaining_ms = int((active_deadline - time.perf_counter()) * 1000)
                        if remaining_ms <= 0:
                            break
                        ready_count = safe_poll(poller, poll_events, max(1, remaining_ms))
                        if not ready_count:
                            continue
                        for offset in range(ready_count):
                            index = poll_events.slot(offset)
                            if index < 0 or index >= len(sockets):
                                continue
                            current_sock = sockets[index]
                            ev = poll_events.revents(offset)
                            if (
                                ev & int(zlink.PollEventFlag.POLLOUT)
                                and not waiting_reply[index]
                                and send_pending[index]
                            ):
                                seq += 1
                                payload = stamp_payload(
                                    payloads[index],
                                    phase=1,
                                    run_id=run_id,
                                    seq=seq,
                                )
                                if send_nonblocking(
                                    current_sock,
                                    payload,
                                    routing_id=b"SERVER",
                                ):
                                    waiting_reply[index] = True
                                    send_pending[index] = False
                            if not (ev & int(zlink.PollEventFlag.POLLIN)):
                                continue
                            while True:
                                msg = recv_nonblocking(
                                    current_sock,
                                    storage=recv_storage[index],
                                )
                                if msg is None:
                                    break
                                with msg:
                                    data = received_metric_payload(msg)
                                    lat = (
                                        latency_ns_from_message(data)
                                        if data
                                        else None
                                    )
                                    # C: every matched header counts; latency
                                    # excludes clock-skew, halved for round trip.
                                    if data and is_active_message(
                                        data,
                                        expected_msg_size=args.msg_size,
                                        run_id=run_id,
                                    ):
                                        received += 1
                                        if lat is not None:
                                            latencies.append(lat / 2.0)
                                waiting_reply[index] = False
                                if time.perf_counter() < active_deadline:
                                    send_pending[index] = True
                if received == 0:
                    raise RuntimeError(
                        "multi router-router benchmark did not receive any active reply"
                    )
                metrics = result_metrics(
                    count=received,
                    msg_size=args.msg_size,
                    elapsed_s=args.duration,
                    latencies_ns=latencies,
                    bandwidth_multiplier=2.0,
                )
                print_result_lines(
                    "MULTI_ROUTER_ROUTER", args.transport, args.msg_size, metrics
                )
        finally:
            for sock in sockets:
                try:
                    sock.close()
                except Exception as exc:
                    print(f"[perf] close failed: {exc}", file=sys.stderr)


if __name__ == "__main__":
    main()

File: bindings/rust/perf/multi/src/perf_multi_router_router_client.rs
#[path = "perf_common.rs"]
mod common;

use std::time::{Duration, Instant};
use zlink::{
    Message, POLLIN, POLLOUT, PollEvent, Poller, RecvFlags, RouterSocket, RoutingId, SocketMonitor,
    SubmitResult,
};

fn drain_socket(
    index: usize,
    socket: &RouterSocket,
    msg_size: usize,
    latency: &mut common::LatencyStats,
    waiting_reply: &mut [bool],
) -> bool {
    let mut processed = false;
    let mut received = zlink::Received::empty();
    loop {
        match socket.recv(&mut received, RecvFlags::DONT_WAIT) {
            Ok(true) => {
                let data = common::message_payload(received.parts());
                if !common::is_valid_active_message(data, msg_size) {
                    continue;
                }
                let sent_ts_ns = common::decode_sent_ts_ns(data);
                let latency_ns =
                    common::now_ns().saturating_sub(sent_ts_ns.max(0) as u64) as f64 / 2.0;
                latency.record_ns(latency_ns);
                waiting_reply[index] = false;
                processed = true;
            }
            Ok(false) => break,
            Err(err) => panic!("recv failed: {err}"),
        }
    }
    processed
}

fn main() {
    let args = common::MultiArgs::parse();
    let settings = common::MultiSettings::from_env();

    let ctx = common::perf_client_context();
    let server_rid = RoutingId::from(b"perf-rr-server");
    let mut sockets: Vec<RouterSocket> = Vec::with_capacity(settings.clients);
    let payload_size = args.msg_size.max(common::HEADER_SIZE);
    let mut waiting_reply = vec![false; settings.clients];
    let mut send_pending = vec![false; settings.clients];
    let mut seqs = vec![1u64; settings.clients];
    let mut monitors: Vec<SocketMonitor> = Vec::with_capacity(settings.clients);

    for index in 0..settings.clients {
        let sock = ctx.router_socket().expect("router");
        sock.common_options()
            .set_send_high_water_mark(settings.send_high_water_mark)
            .expect("sndhwm");
        sock.common_options()
            .set_receive_high_water_mark(settings.receive_high_water_mark)
            .expect("rcvhwm");
        sock.common_options()
            .set_receive_timeout(Duration::from_millis(1))
            .expect("recv timeout");
        let rid = RoutingId::from(format!("CLIENT-{index}").as_bytes());
        sock.set_routing_id(&rid).expect("set rid");
        sock.router_options()
            .set_connect_routing_id(&server_rid)
            .expect("connect rid");
        if matches!(args.transport.as_str(), "tls" | "wss") {
            let tls = common::resolve_perf_tls_paths().expect("TLS certs not found");
            common::setup_raw_tls_client(&sock, &tls).expect("client tls");
        }
        let mon = SocketMonitor::open(&sock).expect("monitor");
        sock.connect(&args.endpoint).expect("connect");
        sockets.push(sock);
        monitors.push(mon);
    }

    let ready_timeout = common::resolve_multi_connect_ready_timeout();
    for mon in &mut monitors {
        common::wait_monitor_ready(mon, ready_timeout, "multi router-router client");
    }

    // Match the dealer-router client: unified poller with signal-driven
    // perf_socket_poll(...,-1) when no socket made progress. The previous
    // hot-loop thread::sleep(1ms) throttled small-message throughput to
    // ~18-21% of C; a poller-driven wait removes that fixed-latency floor.
    let poller = Poller::new().expect("poller");
    for (index, sock) in sockets.iter().enumerate() {
        poller
            .add_socket(sock, POLLIN | POLLOUT, index)
            .expect("poller add");
    }
    let mut poll_events = vec![PollEvent::default(); sockets.len().max(1)];

    let mut latency = common::LatencyStats::new();
    let deadline = Instant::now() + Duration::from_secs(settings.duration_seconds);
    while Instant::now() < deadline {
        let mut progressed = false;
        for index in 0..sockets.len() {
            progressed |= drain_socket(
                index,
                &sockets[index],
                args.msg_size,
                &mut latency,
                &mut waiting_reply,
            );
            if waiting_reply[index] {
                continue;
            }
            let mut msg = Message::with_size(payload_size).expect("msg");
            common::encode_header(
                msg.data_mut(),
                common::PHASE_ACTIVE,
                args.msg_size as u32,
                seqs[index],
            );
            match perf_submit_measurement!(sockets[index].send(&server_rid), msg) {
                Ok(()) => {
                    waiting_reply[index] = true;
                    send_pending[index] = false;
                    seqs[index] += 1;
                    progressed = true;
                }
                Err(err) if err.code() == SubmitResult::Backpressured => {
                    send_pending[index] = true;
                }
                Err(err) => panic!("send failed: {err}"),
            }
        }
        if progressed {
            continue;
        }
        if Instant::now() >= deadline {
            break;
        }
        let remaining_ms = deadline
            .saturating_duration_since(Instant::now())
            .as_millis()
            .max(1) as i64;
        match poller.wait(&mut poll_events, remaining_ms) {
            Ok(_) => {}
            Err(err) => panic!("poller wait failed: {err}"),
        }
    }

    common::print_result(
        "MULTI_ROUTER_ROUTER",
        &args.transport,
        args.msg_size,
        settings.duration_seconds,
        &latency.finish(),
    );
}
